keras/utils/metrics_utils.py

```diff
@@ -174,7 +174,13 @@ def merge_fn_wrapper(distribution, merge_fn, *args):
                 # Wrapping result in identity so that control dependency between
                 # update_op from `update_state` and result works in case result
                 # returns a tensor.
-                return tf.identity(result)
+                if isinstance(result, dict):
```

Comment:
I think the logic between L140 and L157 can be reuse here, how about extract that into a function and reuse it?